File: executor/environments/typescript/runner.ts
//USER_CODE_PLACEHOLDER

import fs from "fs";
import util from "util";

declare global {
  function solution(args: Record<string, Value>): any;
}

interface Value {
  type: string;
  value: string;
}

interface TestSuite {
  problem_id: string;
  tests: {
    inputs: Record<string, Value>;
    expected: Value;
  }[];
}

interface TestCaseResult {
  success: boolean;
  stdout: string;
  stderr: string;
  runtimeMs: string;
  assertionResults: {
    expected: any;
    received: any;
  }[];
}

interface TestSuiteResult {
  success: boolean;
  message: string;
  passed: number;
  failed: number;
  testResults: TestCaseResult[];
}

function parseValue(input: Value): any {
  switch (input.type) {
    case "string":
      return input.value;
    case "number":
    case "int":
      return Number(input.value);
    case "boolean":
      return input.value === "true";
    case "array":
      return JSON.parse(input.value);
    default:
      throw new Error("Unknown type");
  }
}

function compareValues(expected: Value, received: any): boolean {
  switch (expected.type) {
    case "string":
      return received === expected.value;
    case "number":
      return received === Number(expected.value);
    case "boolean":
      return received === (expected.value === "true");
    case "array":
      return JSON.stringify(received) === expected.value;
    default:
      throw new Error("Unknown type");
  }
}

function measureTimeMs(cb: Function, ...args: any[]) {
  const start = process.hrtime.bigint();
  const result = cb(...args);
  const end = process.hrtime.bigint();

  return {
    result,
    timeMs: Number(end - start) / 1e6,
  };
}

function run() {
  const testSuiteResult: TestSuiteResult = {
    success: true,
    message: "",
    passed: 0,
    failed: 0,
    testResults: [],
  };
  try {
    const args = process.argv.slice(2);
    const testsPath = args[0];
    const testSuite = JSON.parse(
      fs.readFileSync(testsPath, "utf8")
    ) as TestSuite;

    testSuite.tests.forEach((test) => {
      let testStdout = "";
      let testStderr = "";

      // Capture stdout
      const originalStdoutWrite = process.stdout.write.bind(process.stdout);
      // @ts-ignore
      process.stdout.write = (chunk) => {
        testStdout += chunk;
      };

      // Capture stderr
      const originalStderrWrite = process.stderr.write.bind(process.stderr);
      // @ts-ignore
      process.stderr.write = (chunk) => {
        testStderr += chunk;
      };

      // Capture console.log and console.error
      const originalConsoleLog = console.log;
      const originalConsoleError = console.error;
      console.log = (...args) => {
        const output = util.format(...args) + "\n";
        testStdout += output;
      };
      console.error = (...args) => {
        const output = util.format(...args) + "\n";
        testStderr += output;
      };

      const expected = test.expected.value;

      Object.entries(test.inputs).map(([key, value]) => {
        test.inputs[key] = parseValue(value);
      });

      const { result, timeMs } = measureTimeMs(solution, test.inputs);

      process.stdout.write = originalStdoutWrite;
      process.stderr.write = originalStderrWrite;
      console.log = originalConsoleLog;
      console.error = originalConsoleError;

      const testResult: TestCaseResult = {
        success: compareValues(test.expected, result),
        stdout: testStdout,
        stderr: testStderr,
        runtimeMs: timeMs.toString(),
        assertionResults: [
          {
            expected,
            received: JSON.stringify(result) || "null",
          },
        ],
      };

      if (testResult.success) {
        testSuiteResult.passed++;
      } else {
        testSuiteResult.failed++;
      }

      testSuiteResult.testResults.push(testResult);
    });

    testSuiteResult.success = testSuiteResult.failed === 0;
  } catch (error) {
    testSuiteResult.success = false;
    if (error instanceof Error) {
      testSuiteResult.message = error.message;
    } else {
      testSuiteResult.message = "Unknown error";
    }
  } finally {
    fs.writeFileSync("test-results.json", JSON.stringify(testSuiteResult));
  }
}

run();

File: executor/environments/python/runner.py
import json
import sys
import time
from typing import Any, Dict, List, Tuple
from io import StringIO


class Value:
    def __init__(self, type: str, value: str):
        self.type = type
        self.value = value


class TestSuite:
    def __init__(self, problemId: str, tests: List[Dict[str, Any]]):
        self.problemId = problemId
        self.tests = tests


class TestCaseResult:
    def __init__(
        self,
        success: bool,
        stdout: str,
        stderr: str,
        runtimeMs: str,
        assertionResults: List[Dict[str, Any]],
    ):
        self.success = success
        self.stdout = stdout
        self.stderr = stderr
        self.runtimeMs = runtimeMs
        self.assertionResults = assertionResults


class TestSuiteResult:
    def __init__(
        self,
        success: bool,
        message: str,
        passed: int,
        failed: int,
        testResults: List[TestCaseResult],
    ):
        self.success = success
        self.message = message
        self.passed = passed
        self.failed = failed
        self.testResults = testResults


def parse_value(input: Value) -> Any:
    if input.type == "string":
        return input.value
    elif input.type == "int":
        return int(input.value)
    elif input.type == "boolean": 
        return input.value.lower() == "true"
    elif input.type.startswith("array"):
        return json.loads(input.value)
    else:
        raise ValueError("Unknown type")


def compare_values(expected: Value, received: Any) -> bool:
    if expected.type == "string":
        return received == expected.value
    elif expected.type == "number" or expected.type == "int":
        return received == float(expected.value)
    elif expected.type == "boolean":
        return received == (expected.value.lower() == "true")
    elif expected.type.startswith("array"):
        return json.dumps(received, separators=(',', ':')) == expected.value
    else:
        raise ValueError("Unknown type")


def measure_time_ms(cb: callable, *args) -> Dict[str, Any]:
    start = time.perf_counter_ns()
    result = cb(*args)
    end = time.perf_counter_ns()

    return {"result": result, "time_ms": (end - start) / 1e6}


def run():
    test_suite_result = TestSuiteResult(
        success=True, message="", passed=0, failed=0, testResults=[]
    )

    try:
        args = sys.argv[1:]
        tests_path = args[0]

        with open(tests_path, "r") as f:
            test_suite = TestSuite(**json.load(f))

        for test in test_suite.tests:
            stdout_capture = StringIO()
            stderr_capture = StringIO()

            # sys.stdout = stdout_capture
            # sys.stderr = stderr_capture

            expected = test["expected"]["value"]
            inputs = {k: parse_value(Value(**v)) for k, v in test["inputs"].items()}
            print(inputs)

            measured = measure_time_ms(solution, inputs)
            result = measured["result"]
            time_ms = measured["time_ms"]

            sys.stdout = sys.__stdout__
            sys.stderr = sys.__stderr__

            test_result = TestCaseResult(
                success=compare_values(Value(**test["expected"]), result),
                stdout=stdout_capture.getvalue(),
                stderr=stderr_capture.getvalue(),
                runtimeMs=str(time_ms),
                assertionResults=[
                    {
                        "expected": expected,
                        "received": (
                            json.dumps(result, separators=(',', ':')).strip('"') if result is not None else "null"
                        ),
                    }
                ],
            )

            if test_result.success:
                test_suite_result.passed += 1
            else:
                test_suite_result.failed += 1

            test_suite_result.testResults.append(vars(test_result))

        test_suite_result.success = test_suite_result.failed == 0
    except Exception as error:
        test_suite_result.success = False
        test_suite_result.message = str(error)
    finally:
        with open("test-results.json", "w") as f:
            json.dump(test_suite_result.__dict__, f)


if __name__ == "__main__":
    run()
